### PR TITLE
Add option to disable SSL verification in requests

### PR DESCRIPTION
**Pull request by k3an3**
This PR adds the ability to disable SSL verification for requests against servers with an invalid SSL configuration or self-signed certificates.

I began writing some tests, although I am looking for suggestions on how to proceed with them in the cleanest way. Right now, they are a bit broken.

### CODE REVIEW

**Comment by inglesp** [comment]
Hi

**Comment by inglesp** [comment]
Hi @keaneokelley 

I have no idea what happened to my comment in March!

Thanks for this PR -- it's a good addition.

As for testing, I think all you need to test is that in the case that a server has a bad certificate, requesting a page over SSL fails when `verify=True` and succeeds when `verify=False` -- so you don't need to serve the whole of the `site/` directory over SSL (that way, I think you'll have problems with the `http://` in the links).

Instead, you should create a new directory at `tests/one-page-site/` containing just an `index.html` that has no links.

Then, in `serve()` you should change `_serve()` so that it takes an `bad_ssl_cert` argument which can be `True` or `False` (the default).  If `False`, it should do what we currently do; if `True` it should do the `ssl.wrap_socket()` stuff, with a bad SSL certificate.

`serve()` should then start another process on port 8002, serving the `one-page-site` site with a bad SSL certificate.  You can then test that `verify=True` / `verify=False` does what you expect.

Does that make sense?

**Comment by k3an3** [comment]
Thanks for the suggestions, I think I was able to implement it.

Only problem, the tests for Python 3.4 are not passing and I've not figured out why.

File: tests/test_http_crawler.py [hunk 1]
```diff
@@ -159,3 +176,22 @@ def test_extract_urls_from_css():
...
+
+def test_ssl_verify_true():
+    serve()
+
+    e = None
+    try:
+        list(http_crawler.crawl('https://localhost:8002', verify=True))
+    except SSLError:
+        e = True
+    assert e
```
Comment (inglesp) [inline tests/test_http_crawler.py]: You should use [`pytest.raises`](http://pytest.readthedocs.io/en/reorganize-docs/new-docs/user/pytest_raises.html) here.

**Review by inglesp — CHANGES_REQUESTED** [review]
Thanks!

I've left one comment, and I haven't dug into the test failure.  I'm about to go on holiday for a few days, but will investigate once I'm back if you haven't managed to track down the problem.

**Comment by inglesp** [comment]
Thanks for the update.

The problem with the tests is that `ssl.PROTOCOL_TLS` is not defined in Python 3.4.  I suggest dropping the `ssl_version` argument and seeing what happens -- it looks like Python provides a sensible default and doing so works for 3.4 and 3.6.

(If you're interested, I managed to work this out by running just the failing test with `python -m pytest tests/test_http_crawler.py::test_ssl_verify_false` and this included a helpful traceback in the output.)

**Comment by k3an3** [comment]
Should be passing now. The only thing that is failing (coverage-report) fails on master as well.

**Comment by inglesp** [comment]
The failing tests were (I think) caused by flakiness in the tests themselves, and re-running them has turned Travis green.  (This may also be why the cause of failures on 3.4 weren't obvious.)

Thanks for your work on this!

**Review by inglesp — APPROVED** [review]

**Pull request merged**